icicle_v3/backend/cpu/src/hash/poseidon/cpu_poseidon.h

```diff
@@ -0,0 +1,203 @@
+#include "hash.h"                   // DEBUG - use common hash.h and not from the local dir.
+#include "icicle/fields/field_config.h"
+
+// typedef uint32_t limb_t;            // DEBUG. To remove after Hadar submit hash.h.
+
+namespace icicle {
+ 
+/**
+ * @brief Poseidon hash constant struct.
+ */
+template <typename S>
+struct PoseidonConstants {
+  const S* m_rounds_constants;     ///< Rounds constants of this Poseidon hash.
+  const S* m_mds_matrix;           ///< Full rounds MDS matrix.
+  const S* m_pre_matrix;           ///< Pre-matrix (used at the last round of the upper full rounds).
+  const S* m_sparse_round_matrix;  ///< Partial rounds MDS matrix.
+  PoseidonConstants(const S* m_rounds_constants, const S* m_mds_matrix, const S* m_pre_matrix, const S* m_sparse_round_matrix) :
+    m_rounds_constants(m_rounds_constants), m_mds_matrix(m_mds_matrix),
+    m_pre_matrix(m_pre_matrix), m_sparse_round_matrix(m_sparse_round_matrix) {}
+};
+
+/**
+ * @brief Poseidon hash class.
+ */
+// template <typename S = scalar_t>
+template <typename S>
+class Poseidon : public Hash {
+  public:
+    Poseidon (
+      const unsigned int  arity,
+      const unsigned int  alpha,
+      const unsigned int  nof_partial_rounds,
+      const unsigned int  nof_upper_full_rounds,
+      const unsigned int  nof_end_full_rounds,
+      const S*            rounds_constants,
+      const S*            mds_matrix,
+      const S*            pre_matrix,
+      const S*            sparse_matrix
+    );
+
+    /**
+     * @brief Function to run a single Poseidon hash.
+     * @param input_limbs Pointer to the Poseidon hash input limbs.
+     * @param output_limbs Pointer to the Poseidon hash output limbs.
+     * @return Error code of type eIcicleError.
+     */
+    eIcicleError run_single_hash (const limb_t *input_limbs, limb_t *output_limbs, const HashConfig& config) const;
+
+    /**
+     * @brief Function to run multiple Poseidon hashes.
+     * @param input_limbs Pointer to the input limbs.
+     * @param output_limbs Pointer to the output limbs.
+     * @param nof_hashes Number of hashes to run.
+     * @return Error code of type eIcicleError.
+     */
+    eIcicleError run_multiple_hash(const limb_t *input_limbs, limb_t *output_limbs, int nof_hashes, const HashConfig& config, const limb_t *side_input_limbs = nullptr) const;
+
+  private:
+    const unsigned int    m_arity;     ///< m_arity of this Poseidon hash.
+    const unsigned int    m_alpha;     ///< m_alpha of this Poseidon hash S-box.
+    const unsigned int    m_nof_partial_rounds;     ///< Partial number of rounds of this Poseidon hash.
+    const unsigned int    m_nof_upper_full_rounds;     ///< Number of full rounds at the beginning of this Poseidon hash.
+    const unsigned int    m_nof_end_full_rounds;     ///< Number of full rounds at the end of this Poseidon hash.
+    PoseidonConstants<S>  m_poseidon_constants;      ///< Structure that holds Poseidon hash round constants and MDS matrix values.
```

Comment:
Isn't it going to be more comfortable to just save the 4 vectors instead of this class?

Reply:
What do you mean? What class?